backend/src/main/java/gov/cdc/usds/simplereport/api/patientExperience/PatientExperienceController.java

```diff
@@ -0,0 +1,117 @@
+import static gov.cdc.usds.simplereport.api.Translators.parsePhoneNumber;
+import static gov.cdc.usds.simplereport.api.Translators.parseRace;
+import static gov.cdc.usds.simplereport.api.Translators.parseState;
+import static gov.cdc.usds.simplereport.api.Translators.parseString;
+
+import gov.cdc.usds.simplereport.api.exceptions.FeatureFlagDisabledException;
+import gov.cdc.usds.simplereport.api.exceptions.InvalidPatientLinkException;
+import gov.cdc.usds.simplereport.api.model.AoEQuestions;
+import gov.cdc.usds.simplereport.api.model.pxp.PxpApiWrapper;
+import gov.cdc.usds.simplereport.api.model.pxp.PxpPersonWrapper;
+import gov.cdc.usds.simplereport.db.model.Organization;
+import gov.cdc.usds.simplereport.db.model.PatientLink;
+import gov.cdc.usds.simplereport.db.model.Person;
+import gov.cdc.usds.simplereport.db.model.TestEvent;
+import gov.cdc.usds.simplereport.db.model.auxiliary.TestResult;
+import gov.cdc.usds.simplereport.service.PatientLinkService;
+import gov.cdc.usds.simplereport.service.PersonService;
+import gov.cdc.usds.simplereport.service.TestEventService;
+import gov.cdc.usds.simplereport.service.TestOrderService;
+
+@RestController
+@RequestMapping(value = "/pxp")
+@Validated
+public class PatientExperienceController {
+  private static final Logger LOG = LoggerFactory.getLogger(PatientExperienceController.class);
+
+  @Autowired
+  private PersonService ps;
+
+  @Autowired
+  private PatientLinkService pls;
+
+  @Autowired
+  private TestOrderService tos;
+
+  @Autowired
+  private TestEventService tes;
+
+  @Value("${feature-flags.patient-links:false}")
+  private boolean patientLinksEnabled;
+
+  /**
+   * Verify that the patient-provided DOB matches the patient on file for the
+   * patient link id. It returns the full patient object if so, otherwise it
+   * throws an exception
+   */
+  @PutMapping("/link/verify")
+  public PxpPersonWrapper getPatientLinkVerify(@RequestBody PxpApiWrapper<Void> body)
+      throws InvalidPatientLinkException {
+    if (!patientLinksEnabled) {
+      throw new FeatureFlagDisabledException("Patient links not enabled");
+    }
+    Person p = pls.getPatientLinkVerify(body.getPlid(), body.getDob());
+    TestEvent te = tes.getLastTestResultsForPatient(p);
+    return new PxpPersonWrapper(p, te);
+  }
+
+  @PutMapping("/patient")
+  public Person updatePatient(@RequestBody PxpApiWrapper<Person> body) throws InvalidPatientLinkException {
+    if (!patientLinksEnabled) {
+      throw new FeatureFlagDisabledException("Patient links not enabled");
+    }
+    pls.getPatientLinkVerify(body.getPlid(), body.getDob()); // gotta verify!
+
+    PatientLink pl = pls.getPatientLink(body.getPlid());
+    UUID facilityId = pl.getTestOrder().getFacility().getInternalId();
+    String patientId = pls.getPatientLinkVerify(body.getPlid(), body.getDob()).getInternalId().toString();
+    Person person = body.getData();
+
+    return ps.updatePatient(facilityId, patientId, parseString(person.getLookupId()),
```

Comment:
I hate this service method so much.

No, I don't have a particularly good plan for fixing it right now, I just wanted you to know that you weren't the only one.

Reply:
Haha, thanks, I'm glad to know I'm not alone